guides/v2.3/graphql/develop/resolvers.md

```diff
@@ -42,7 +15,185 @@ $info | [`Magento\Framework\GraphQl\Schema\Type\ResolveInfo`]({{ site.mage2blobu
+A GraphQL resolver must implement one of the following interfaces:
+
+-  [`\Magento\Framework\GraphQl\Query\Resolver\BatchResolverInterface`]({{ site.mage2bloburl }}/{{ page.guide_version }}/lib/internal/Magento/Framework/GraphQl/Query/Resolver/BatchResolverInterface.php)
+
+-  [`\Magento\Framework\GraphQl\Query\Resolver\BatchServiceContractResolverInterface`]({{ site.mage2bloburl }}/{{ page.guide_version }}/lib/internal/Magento/Framework/GraphQl/Query/Resolver/BatchServiceContractResolverInterface.php)
+
+-  [`\Magento\Framework\GraphQl\Query\ResolverInterface`]({{ site.mage2bloburl }}/{{ page.guide_version }}/lib/internal/Magento/Framework/GraphQl/Query/ResolverInterface.php)
+
+The first two interfaces provide a way to resolve multiple branches/leaves at once (known as batching), while the last one resolves one request at a time. Magento recommends using batch resolvers for queries because they improve performance by fetching information required to resolve multiple GraphQL requests with a single operation.
+
+## Query resolvers
+
+### BatchResolverInterface
+
+Batch resolvers gather GraphQL requests for the same field until there is no way to process the tree further without resolving previous requests.
+
+Consider the following example:
+
+```graphql
+query ($filter: ProductAttributeFilterInput!) {
+  products (filter: $filter) {
+    items {
+      id
+      sku
+      related_products {
+        sku
+        related_products {
+          sku
+        }
+      }
+    }
+    total_count
+  }
+}
+```
+
+The query loads a list of products, the SKUs of their related products, and then any secondary related product SKUs.
+
+Loading a list of related products individually for each product would be expensive performance-wise. With batch resolvers, you can load linked products for all of products that were initially found, and then group them by root products. After the `items` branch is resolved, a batch resolver for `related_products` will be called for the first product found. But instead of resolving the list right away, it will just add the first product to the list of products that require loading additional related products. After all the products from the `items` branch have been loaded, the lists of related products must be loaded. Then `BatchResolverInterface::resolve()` executes with a gathered list of previous requests to `related_products` branches. At this point, the resolver is able to extract product DTOs from each GraphQL request, load all the product links, sort them by root products, and generate GraphQL values for each branch. After this is done, the same batching will take place when resolving child `related_products` branches.
```

Comment:
```suggestion
Loading a list of related products individually for each product would be expensive performance-wise. With batch resolvers, you can load linked products for all products that were initially found, then group them by root products. After the `items` branch is resolved, a batch resolver for `related_products` will be called for the first product found. Instead of resolving the list right away, it will just add the first product to the list of products that require loading additional related products. After all the products from the `items` branch have been loaded, the lists of related products must be loaded. Then, `BatchResolverInterface::resolve()` executes with a gathered list of previous requests to `related_products` branches. At this point, the resolver is able to extract product DTOs from each GraphQL request, load all the product links, sort them by root products, and generate GraphQL values for each branch. After this is done, the same batching will take place when resolving child `related_products` branches.
```